patches/api/0370-More-Projectile-API.patch

```diff
@@ -136,6 +177,120 @@ index d1b37530319f6d37ee37f62080289c1e45848bc8..e94c7e279356c510f60508b26277d489
 +    void setWaitTime(int ticks);
 +    // Paper end
  }
+diff --git a/src/main/java/org/bukkit/entity/Projectile.java b/src/main/java/org/bukkit/entity/Projectile.java
+index a523fca4baab447181ef91df67fa69b24e010149..53f2c21e1cd07df0a246cc53d0af919aa26f8b46 100644
+--- a/src/main/java/org/bukkit/entity/Projectile.java
++++ b/src/main/java/org/bukkit/entity/Projectile.java
+@@ -43,4 +43,45 @@ public interface Projectile extends Entity {
+      */
+     @Deprecated(forRemoval = true) // Paper
+     public void setBounce(boolean doesBounce);
++    // Paper start
++
++    /**
++     * Gets whether the projectile has left the
++     * hitbox of their shooter and can now hit entities.
++     *
++     * @return has left shooter's hitbox
++     */
++    boolean hasLeftShooter();
++
++    /**
++     * Sets whether the projectile has left the
++     * hitbox of their shooter and can now hit entities.
++     *
++     * This is recalculated each tick if the projectile has a shooter.
++     *
++     * @param leftShooter has left shooter's hitbox
++     */
++    void setHasLeftShooter(boolean leftShooter);
++
++    /**
++     * Gets whether the projectile has been
++     * shot into the world and has sent a projectile
++     * shot game event.
++     *
++     * @return has been shot into the world
++     */
++    boolean hasBeenShot();
++
++    /**
++     * Sets whether the projectile has been
++     * shot into the world and has sent a projectile
++     * shot game event.
++     *
++     * Setting this to false will cause a game event
++     * to fire and the value to be set back to true.
++     *
++     * @param beenShot has been in shot into the world
++     */
++    void setHasBeenShot(boolean beenShot);
++    // Paper end
+ }
+diff --git a/src/main/java/org/bukkit/entity/ShulkerBullet.java b/src/main/java/org/bukkit/entity/ShulkerBullet.java
+index 4623e0d767b343cbdc6fcf20b3b2ff7ff14863cf..00e63e8df54cd6a3cbe0ca8578c6e74e6ab0223d 100644
+--- a/src/main/java/org/bukkit/entity/ShulkerBullet.java
++++ b/src/main/java/org/bukkit/entity/ShulkerBullet.java
+@@ -18,4 +18,59 @@ public interface ShulkerBullet extends Projectile {
+      * @param target the entity to target
+      */
+     void setTarget(@Nullable Entity target);
++
++    /**
++     * Gets where the shulker bullet should move towards.
++     * This is usually relative towards their target.
++     *
++     * @return target
++     */
++    @org.jetbrains.annotations.NotNull
++    org.bukkit.util.Vector getTargetDelta();
```

Comment:
You might want to elaborate more on what a target delta here is, specifically in the shulker bullet case.
This changed on a tick base, the target delta setter should reference that.
It also isn't really the final destination, I think that would be good to reference here.